### PR TITLE
added final parser implementation

### PR DESCRIPTION
**Pull request by IuliaPapureanu**

### CODE REVIEW

File: lab5-8/.idea/.gitignore [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# Default ignored files
```
Comment (Patrulescu-Ronald-Sandrino) [inline lab5-8/.idea/.gitignore:1]: add `.idea/` & `**/_pycache__`, or even better, use a .gitignore template

Comment (Patrulescu-Ronald-Sandrino) [inline lab5-8/.idea/.gitignore:1]: Directories like `.idea` & `__pycache_` have no utility in the source code repository. It makes it harder for me to review your code and I suppose they are also a pain for the developer. 
Suggestion: use .gitignore templates (like the ones from https://github.com/github/gitignore).
They also have [a nice .gitignore file for Python code](https://github.com/github/gitignore/blob/main/Python.gitignore)

File: lab5-8/main.py [hunk 2]
```diff
@@ -0,0 +1,4 @@
+from ui import UI
+
+ui = UI()
```
Comment (Patrulescu-Ronald-Sandrino) [inline lab5-8/main.py:3]: you can also `UI().run()`

**Review by birthleft — COMMENTED** [review]
Configuration and Build-related files seem to have been committed as well.

A suggestion would be using .gitignore to mark these files as irrelevant to the source code.

If you are using PyCharm as an IDE, you can use a plugin known as "[.ignore](https://plugins.jetbrains.com/plugin/7495--ignore)", which can help you out in adding the right files to the ".gitignore" and even offer you base templates.

**Review by birthleft — COMMENTED** [review]
Personal preference, but a few more comments would have been a very pleasant addition. Not a dealbreaker, but something to keep in mind in the future. :)

**Review by birthleft — COMMENTED** [review]
Everything else seems alright. Due to Python's syntax rules, the code is well-spaced, not too clustered, which discourages the possibility of "spaghetti code" appearing within source files.  The existence of the documentation within the GitHub project is also a really neat addition.

Other than that, yeah, pretty well-written code.

**Review by birthleft — APPROVED** [review]

File: lab5-8/domain/grammar.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    @staticmethod
+    def fromFile(fileName):
+
+        with open(fileName, 'r') as file:
+            N = Grammar.parseLine(file.readline())
+            E = Grammar.parseLine(file.readline())
+            S = file.readline().split('=')[1].strip()
+            P = Grammar.parseRules(Grammar.parseLine(''.join([line for line in file])))
+
+            #if not Grammar.validate(N, E, P, S):
```
Comment (Patrulescu-Ronald-Sandrino) [inline lab5-8/domain/grammar.py:36]: if you need to frequently comment out this part, consider using a boolean switch

File: lab5-8/domain/grammar.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+                    result[lhs] = [(value,index)]
+                index+=1
+
+        return result
+
+    def splitRhs(self, prod):
+        return prod.split(' ')
+
+    def isNonTerminal(self, value):
+        return value in self.N
```
Comment (Patrulescu-Ronald-Sandrino) [inline lab5-8/domain/grammar.py:65]: more efficient if `self.N` was a set

File: lab5-8/domain/grammar.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+    def isTerminal(self, value):
+        return value in self.E
```
Comment (Patrulescu-Ronald-Sandrino) [inline lab5-8/domain/grammar.py:68]: more efficient if `self.N` was a set

File: lab5-8/ui.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+        print('Read g2')
+
+    def readSequence(self, fname):
+        sequence = ""
+        with open(fname, 'r') as fin:
+            for line in fin.readlines():
+                sequence += line.strip() + " "
+        return sequence.strip()
+
+    def evaluateG1(self):
```
Comment (Patrulescu-Ronald-Sandrino) [inline lab5-8/ui.py:36]: only the grammar file and the sequence file differ between this method and `evaluateG2`, could have been instead a method with those 2 as parameters

**Review by Patrulescu-Ronald-Sandrino — APPROVED** [review]
Consider using [type hinting](https://realpython.com/lessons/type-hinting/). It makes the code easier to read.

**Pull request merged**